// resedit-cli/dist/requestSimple/requestSimpleUsingFetch.js
import { EnvHttpProxyAgent, setGlobalDispatcher } from 'undici';
import * as log from '../log.js';
const fetchFunction = (() => {
    if (typeof fetch === 'function') {
        if (process.env.HTTPS_PROXY || process.env.HTTP_PROXY) {
            setGlobalDispatcher(new EnvHttpProxyAgent());
        }
        return fetch;
    }
    return null;
})();
export function isAvailable() {
    return fetchFunction !== null;
}
export default function requestSimpleUsingFetch(url, opt, cb) {
    log.debug(`[sign] Use${typeof fetch !== 'undefined' ? ' native' : ''} fetch function`);
    async function inner() {
        try {
            const response = await fetchFunction(url, {
                method: opt.method,
                headers: opt.headers,
                body: opt.body,
            });
            const buffer = Buffer.from(await response.arrayBuffer());
            const cbHeaders = {};
            response.headers.forEach((value, key) => {
                cbHeaders[key] = value;
            });
            if (response.ok) {
                cb(null, cbHeaders, buffer);
            }
            else {
                cb(new Error(`Server error ${response.status} ${response.statusText}`), cbHeaders, buffer);
            }
        }
        catch (error) {
            cb(error, {}, Buffer.from(''));
        }
    }
    inner();
}
